src/app/qgsattributesformproperties.cpp

```diff
@@ -509,43 +511,72 @@ QTreeWidgetItem *QgsAttributesFormProperties::loadAttributeEditorTreeItem( QgsAt
+{
+  disconnect( mFormLayoutTree, &QTreeWidget::itemSelectionChanged, this, &QgsAttributesFormProperties::onFormLayoutSelectionChanged );
+  synchronizeDnDTrees( mAvailableWidgetsTree, mFormLayoutTree );
+  connect( mFormLayoutTree, &QTreeWidget::itemSelectionChanged, this, &QgsAttributesFormProperties::onFormLayoutSelectionChanged );
+}
+
+void QgsAttributesFormProperties::onFormLayoutSelectionChanged()
+{
+  // when the selection changes in the DnD layout, sync the main tree
+  disconnect( mAvailableWidgetsTree, &QTreeWidget::itemSelectionChanged, this, &QgsAttributesFormProperties::onAttributeSelectionChanged );
+  synchronizeDnDTrees( mFormLayoutTree, mAvailableWidgetsTree );
+  connect( mAvailableWidgetsTree, &QTreeWidget::itemSelectionChanged, this, &QgsAttributesFormProperties::onAttributeSelectionChanged );
+}
+
+void QgsAttributesFormProperties::synchronizeDnDTrees( DnDTree *emitter, DnDTree *receiver )
 {
   storeAttributeTypeDialog();
   storeAttributeRelationEdit();
 
-  switch ( mAvailableWidgetsTree->currentItem()->data( 0, DnDTreeRole ).value<DnDTreeItemData>().type() )
+  if ( emitter->selectedItems().count() != 1 )
   {
-    case DnDTreeItemData::Relation:
-    {
-      mAttributeTypeDialog->setVisible( false );
-      loadAttributeRelationEdit();
-      break;
-    }
-    case DnDTreeItemData::Field:
-    {
-      mAttributeRelationEdit->setVisible( false );
-      loadAttributeTypeDialog();
-      break;
-    }
-    case DnDTreeItemData::Container:
-    {
-      mAttributeRelationEdit->setVisible( false );
-      mAttributeTypeDialog->setVisible( false );
-      break;
-    }
-    case DnDTreeItemData::QmlWidget:
-    {
-      mAttributeRelationEdit->setVisible( false );
-      mAttributeTypeDialog->setVisible( false );
-      break;
-    }
-    case DnDTreeItemData::HtmlWidget:
+    clearAttributeTypeFrame();
+    receiver->clearSelection();
+  }
+  else
+  {
+    DnDTreeItemData itemData = emitter->selectedItems().at( 0 )->data( 0, DnDTreeRole ).value<DnDTreeItemData>();
+    switch ( itemData.type() )
     {
-      mAttributeRelationEdit->setVisible( false );
-      mAttributeTypeDialog->setVisible( false );
-      break;
+      case DnDTreeItemData::Relation:
+      {
+        receiver->selectFirstMatchingItem( itemData );
+        loadAttributeRelationEdit();
+        break;
+      }
+      case DnDTreeItemData::Field:
+      {
+        receiver->selectFirstMatchingItem( itemData );
+        loadAttributeTypeDialog();
+        break;
+      }
+      case DnDTreeItemData::Container:
+      case DnDTreeItemData::QmlWidget:
+      case DnDTreeItemData::HtmlWidget:
+      {
+        clearAttributeTypeFrame();
+        receiver->clearSelection();
+        break;
+      }
     }
+  }
+}
 
+void QgsAttributesFormProperties::clearAttributeTypeFrame()
+{
+  if ( mAttributeTypeDialog )
+  {
+    mAttributeTypeFrame->layout()->removeWidget( mAttributeTypeDialog );
+    mAttributeTypeDialog->deleteLater();
```

Comment:
I think, since you clear the `QgsAttributeTypeDialog` (like it should) it's not available anymore here: https://github.com/qgis/QGIS/blob/d1126c3a126e93b3bc28436968e8bc79c5b4583f/src/app/qgsattributesformproperties.cpp#L87 on cancellation of the layer property here:
https://github.com/qgis/QGIS/blob/master/src/app/qgsvectorlayerproperties.cpp#L953

This leads to a crash. It would fix it if you check the `QgsAttributeTypeDialog` for existence in the `init()` but I'm not sure why it has to init the Attribute Form Properties on cancellation at all. 

I'm sure it's not your fault but mine, since I did this two years ago. Tell me if I should investigate more or we can have a call, then I can show you the situation.

Reply:
the crash should be fixed with last commit